source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc

```diff
@@ -107,114 +118,187 @@ class TranscoderImpl : public Transcoder {
-    }
-    break;
-  case envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder::
-      DescriptorSetCase::kProtoDescriptorBin:
-    if (!descriptor_set.ParseFromString(proto_config.proto_descriptor_bin())) {
-      throw EnvoyException("transcoding_filter: Unable to parse proto descriptor");
+  // This was not a requested service, so assume this is a duplicate response and ignore it.
+  if (services_.find(service_requested) == services_.end()) {
+    return;
+  }
+
+  for (const std::string& file_descriptor_bytes:
+      message->file_descriptor_response().file_descriptor_proto()) {
+    if (!file_descriptor_set_.add_file()->ParseFromString(file_descriptor_bytes)) {
+      throw EnvoyException(
+          "transcoding_filter: Unable to parse proto descriptor returned from reflection");
     }
-    break;
-  case envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder::
-      DescriptorSetCase::DESCRIPTOR_SET_NOT_SET:
-    throw EnvoyException("transcoding_filter: descriptor not set");
   }
 
-  for (const auto& file : descriptor_set.file()) {
-    addFileDescriptor(file);
+  // Record that we have already received the response for this service.
+  services_.erase(service_requested);
+
+  // We have received all the protos we have asked for, so inject them into the
+  // proto config and signal to InitManager that loading is complete.
+  if (services_.empty()) {
+    descriptor_pool_builder_->loadFileDescriptorProtos(file_descriptor_set_);
+    init_target_->ready();
   }
+}
 
-  convert_grpc_status_ = proto_config.convert_grpc_status();
-  if (convert_grpc_status_) {
-    addBuiltinSymbolDescriptor("google.protobuf.Any");
-    addBuiltinSymbolDescriptor("google.rpc.Status");
+void AsyncReflectionFetcher::onReceiveTrailingMetadata(Http::ResponseTrailerMapPtr&&) {}
+void AsyncReflectionFetcher::onRemoteClose(Envoy::Grpc::Status::GrpcStatus status, const std::string& message) {
+  if (status != Envoy::Grpc::Status::Ok) {
+    // TODO: Replace with std::string.
+    char errorMessage[1024];
+    snprintf(errorMessage, sizeof(errorMessage),
+        "transcoding_filter: server reflection request failed with status %ld: %s",
+        status,
+        message.c_str());
+    throw EnvoyException(errorMessage);
   }
+}
 
-  type_helper_ = std::make_unique<google::grpc::transcoding::TypeHelper>(
-      Protobuf::util::NewTypeResolverForDescriptorPool(Grpc::Common::typeUrlPrefix(),
-                                                       &descriptor_pool_));
+DescriptorPoolBuilder::DescriptorPoolBuilder(
+        const envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder&
+            proto_config,
+        std::shared_ptr<AsyncReflectionFetcher> async_reflection_fetcher) :
+    proto_config_(proto_config),
+    async_reflection_fetcher_(async_reflection_fetcher) {}
+
+
+void DescriptorPoolBuilder::requestDescriptorPool(JsonTranscoderConfig* json_transcoder_config) {
```

Comment:
The biggest problem is: the relationship between DescriptorPoolBuilder and JsonTranscoderConfig is very messy. It is not clear who owns who. 

Let me see if I can come up something better.